### ISSUE TITLE
Encapsualtion

### ISSUE THREAD

**Issue by FXTi** [FXTi/iocp-wrapper#14]
This is for this code to be integrated with Mio.

Implement wrapper and basic tests first.

Wait for more comments.

**Comment by carllerche** [comment]
The goal is to ship a `wepoll` crate. This crate will eventually be included in the `mio` git repository. For now, it will be developed externally (this repository is fine).

Mio is being refactored so that each platform provides all of the necessary types. This makes it easier to develop per platform support as there are no required dependencies.

The types that must be provided are:

- [`Selector`](https://github.com/tokio-rs/mio/blob/master/src/poll.rs#L214)
- [`Event`](https://github.com/tokio-rs/mio/blob/master/src/event/event.rs#L18)
- [`Events`](https://github.com/tokio-rs/mio/blob/master/src/event/events.rs#L46)
- `Interests` (not yet factored out per platform)
- [`Waker`](https://github.com/tokio-rs/mio/blob/master/src/waker.rs#L80)
- [`TcpStream` and `TcpListener`](https://github.com/tokio-rs/mio/blob/master/src/net/tcp.rs)
- [`UdpSocket`](https://github.com/tokio-rs/mio/blob/master/src/net/udp.rs#L95-L98)

These types need to provide the API used by the mio wrapper. The details of this API are going to change a bit as Mio is updated for the 0.7 release. My guess is that at the end, the per-platform API will probably look almost identical to the Mio public API.

You will probably also have your own `Token` type, which represents the epoll "user data".

I would start by sketching these things out.

**Comment by carllerche** [comment]
To start, you can probably just have a `TcpStream`, a `Selector`, and any other type you need to implement that API.

Your goal will be to get `TcpStream::connect(...)`, Then be able to do `my_selector.register(&my_socket, my_token, my_interests)`.

My guess is that you will want `struct State { ... }` to represent the socket state. Then define:

```rust
struct TcpStream {
    sock: WindowsSocketType,
    state: Arc<State>,
}
```

There are a number of things from the C version of wepoll that we won't need in our sock state. I *believe* we don't need `tree_node` since each socket value tracks its own state.

Also note that the `State` struct will be a bit different than how it is defined in wepoll. In wepoll, [the state is allocated when the port is known](https://github.com/piscisaureus/wepoll/blob/master/src/sock.c#L65). In our case, we will allocate the state struct when the socket is initialized. Then, when the socket is registered with the selector, we will fill in the rest of the info.

**Comment by piscisaureus** [comment]
> There are a number of things from the C version of wepoll that we won't need in our sock state. I _believe_ we don't need `tree_node` since each socket value tracks its own state.

Agreed. (Even if we did allow for looking up mio state for a raw `std::os::windows::raw::SOCKET`, I think we should just use a HashMap for that).
In similar vain, we won't need `queue_node` as the "update queue" can live in de VecDeque.
Rust has all these advanced data structures built in, we might as well use them.

**Comment by FXTi** [comment]
I see wepoll use `sock_update_queue` and `sock_delete_queue` in `port_state` to complete necessary sock state changes
I think it's because socket can't do the update work automatically, so embed them in `port_state` and make port do the job
And Mio can do that job in like `TcpStream`(which represent a socket), so `Selector`(which represent a iocp port) doesn't need `update_queue` and `delete_queue` in this case

Is that correct? @carllerche @piscisaureus

**NO, IT IS NOT CORRECT.**

Take `reregister` as an example, when user call, like `selector.reregister(&tcp_stream)`. Yes, `reregister()` can call `CancelIoEx()` at calling moment. But `reregister` still needs to wait for that to complete and then register the new events. And when that happens, user might be far way from `reregister()`. So it's necessary to keep a VecDeque in `Selector` to track this kind of change of states.
